extensions/guacamole-vault/modules/guacamole-vault-ksm/src/main/java/org/apache/guacamole/vault/ksm/user/KsmDirectoryService.java

```diff
@@ -0,0 +1,241 @@
+    private KsmConfigurationService configurationService;
+
+    /**
+     * All expected fields in the KSM configuration JSON blob.
+     */
+    private static final List<String> EXPECTED_KSM_FIELDS = Arrays.asList(
+            SecretsManager.KEY_HOSTNAME,
+            SecretsManager.KEY_CLIENT_ID,
+            SecretsManager.KEY_PRIVATE_KEY,
+            SecretsManager.KEY_CLIENT_KEY,
+            SecretsManager.KEY_APP_KEY,
+            SecretsManager.KEY_OWNER_PUBLIC_KEY,
+            SecretsManager.KEY_SERVER_PUBIC_KEY_ID
+    );
+
+    /**
+     * Return true if the provided input is a valid base64-encoded string,
+     * false otherwise.
+     *
+     * @param input
+     *     The string to check if base-64 encoded.
+     *
+     * @return
+     *     true if the provided input is a valid base64-encoded string,
+     *     false otherwise.
+     */
+    private static boolean isBase64(String input) {
+
+        try {
+            Base64.getDecoder().decode(input);
+            return true;
+        } catch (IllegalArgumentException e) {
+            return false;
+        }
+    }
+
+    /**
+     * Given an attributes-enabled entity, check for the presence of the
+     * KSM_CONFIGURATION_ATTRIBUTE attribute. If it's set, check if it's a valid
+     * KSM one-time token. If so, attempt to translate it to a base-64-encoded
+     * json KSM config blob, and set it back to the provided entity.
+     * If it's already a KSM config blob, validate it as config blob. If either
+     * validation fails, a GuacamoleException will be thrown.
+     *
+     * @param entity
+     *     The attributes-enable entity for which the KSM configuration
+     *     attribute parsing/validation should be performed.
+     *
+     * @throws GuacamoleException
+     *     If the KSM_CONFIGURATION_ATTRIBUTE is set, but fails to validate as
+     *     either a KSM one-time-token, or a KSM base64-encoded JSON config blob.
+     */
+    public void processAttributes(Attributes entity) throws GuacamoleException {
+
+        // By default, if the KSM config attribute isn't being set, pass the
+        // provided attributes through without any changes
+        Map<String, String> attributes = entity.getAttributes();
+
+        // Get the value of the KSM config attribute in the provided map
+        String ksmConfigValue = attributes.get(
+                KsmAttributeService.KSM_CONFIGURATION_ATTRIBUTE);
+
+        // Check if the attribute is set to a non-empty value
+        if (ksmConfigValue != null && !ksmConfigValue.trim().isEmpty()) {
+
+            // If it's already base-64-encoded, it's a KSM configuration blob,
+            // so validate it immediately
+            if(isBase64(ksmConfigValue)) {
+
+                // Attempt to validate the config as a base64-econded KSM config blob
+                try {
+                    KsmConfig.parseKsmConfig(ksmConfigValue);
+
+                    // If it validates, the entity can be returned as-is
```

Comment:
This doesn't make sense to me:
1. But we're not returning anything.
2. What entity?

Reply:
Good catch on it not returning anything - I'll fix that.

The entity is `entity` - the `Attributes` instance that's passed into `processAttributes()`. I thought that was relatively clear?

Reply:
Is it still unclear?

Reply:
> Good catch on it not returning anything - I'll fix that.
> 
> The entity is `entity` - the `Attributes` instance that's passed into `processAttributes()`. I thought that was relatively clear?

I think I just got hung up on "entity" being how the DB refers to the common base representation of a user / user group.

> Is it still unclear?

I think this looks good now.